### PR TITLE
[Proposal] Wallet file encryption

### PR DESCRIPTION
**Pull request by ZedPea**
So we want to encrypt our wallet file with a password, so *evil* attackers can't nab all our funds.

We also want this to be relatively secure, and relatively easy to implement in multiple languages.

This current implementation is using AES Encryption, with a JSON file format. The exact format of the JSON is not decided at all yet, but that's easy to handle, so I won't discuss it much.

For the AES Encryption, we get our AES Key by performing a SHA256 hash of the password.

For the initialization vector, we generate 128 random bytes. We store this unencrypted, after the magic bytes to verify it's a wallet file.

We want to verify that our wallet file is indeed a wallet file, and that our wallet file has correctly decrypted, i.e. the password is correct.

To do this, we utilize 2 different sets of magic bytes.

The first, is prepended to the encrypted data, and written to the file unencrypted. So, when opening a wallet file, we can simply check if these magic bytes are present, and if they are, the wallet file is indeed a wallet file, and not some random other file.

The second set of magic bytes is used to verify the password is correct. These are prepended to the data to be encrypted. Thus, once the file is unencrypted, the data should begin with these magic bytes. If it does, the password is correct.

So, here are the algorithms to encrypt and decrypt.

```
iswalletfilebytes = [m, a, g, i, c, b, y, t, e, s]

isvalidpasswordbytes = [m, a, g, i, c, b, y, t, e, s]

encrypt(string data, string password, string filename)
{
    file = open(filename)

    file.write(iswalletfilebytes)

    aesKey = sha256(password)

    aesInitializationVector = randomBytes(16)

    file.write(aesInitializationVector)

    dataWithMagicBytes = isvalidpasswordbytes + data

    encryptedData = aes.encrypt(aesKey, aesInitializationVector, dataWithMagicBytes)

    file.write(encrypted)
}

decrypt(string password, string filename)
{
    file = open(filename)

    data = file.readAll()

    if (!data.startswith(iswalletfilebytes))
    {
        error("Not a wallet file")
    }

    data = data.skip(iswalletfilebytes.Length)

    aesKey = sha256(password)

    aesInitializationVector = data.take(16)

    data = data.skip(aesInitializationVector.Length)

    unencryptedData = aes.decrypt(aesKey, aesInitializationVector, dataWithMagicBytes)

    if (!unencryptedData.startswith(isvalidpasswordbytes))
    {
        error("wrong password")
    }

    unencryptedData = unencryptedData.skip(isvalidpasswordbytes.length)

    return unencryptedData
}
```

The base code for this algorithm was taken from CodIsAFish, which was adapted from Editfmah's code. We are OK to use this code, crediting both with an MIT License.

I believe this should be relatively cross platform, as long as the language has the ability to perform SHA256, and AES with variable length keys and initialization vectors/block sizes.

Let me know what you think :)

Edit: Updated with security advisement's by elvira - Thanks!

### CODE REVIEW

File: CantiLib/Blockchain/Crypto/JSONWalletEncrypter.cs [hunk 1]
```diff
@@ -14,18 +14,18 @@
 /* Clashes with System.Text Encoding namespace */
 using Data = Canti.Data;
 using Canti.Utilities;
-using Canti.Blockchain.Wallet;
+using Canti.Blockchain.WalletBackend;
 
 namespace Canti.Blockchain.Crypto
 {
     /* So using the Wallet class doesn't conflict with the Wallet namespace */
```
Comment (SoreGums) [inline CantiLib/Blockchain/Crypto/JSONWalletEncrypter.cs]: So using the **WalletBackend** class
?

**Comment by SoreGums** [comment]
That's a pretty elegant solution - nice one.
Especially the whole built-in check to verify the file is a wallet - going to help greatly with UX/UI for the user.

File: CantiLib/Blockchain/Crypto/JSONWalletEncrypter.cs [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+using Newtonsoft.Json;
+
+/* Clashes with System.Text Encoding namespace */
+using Data = Canti.Data;
+using Canti.Utilities;
+using Canti.Blockchain.WalletBackend;
+
+namespace Canti.Blockchain.Crypto
+{
+    /* So using the Wallet class doesn't conflict with the Wallet namespace */
```
Comment (SoreGums) [inline CantiLib/Blockchain/Crypto/JSONWalletEncrypter.cs]: So using the **WalletBackend** class
?

Comment (ZedPea) [reply CantiLib/Blockchain/Crypto/JSONWalletEncrypter.cs]: Yep, forgot to rename after changing wallet to walletbackend, thanks.

**Comment by ZedPea** [comment]
Gonna go ahead and merge since there have been no disagreements with the updated version, and I want to push up more code.

**Pull request merged**